rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,394 @@
+tedious UI navigation system. However, it still asks them to manage the
+graphical aspect of UI such as visual highlights, a more holistic approach
+probably solves this, at the cost of being more complex and less flexible. Such
+a system might be explored in the future, but for the sake of
+having an implementation to work with, I think aiming for a smaller scope is a
+better idea.
+
+### As plugin or integrated in the engine?
+
+I want to first draft this proposition as a standalone plugin ([see on
+github](https://github.com/nicopap/ui-navigation)). The current
+design doesn't require any specific insight into the engine that cannot be
+accessed by a 3rd party plugin. However as I already mentioned, this solves 
+[bevy-specific concerns](https://github.com/bevyengine/bevy/issues/254), and
+bevy would benefit from integrating a good ui navigation system.
+
+### Naming
+
+I'm not comfortable with the names I chose for the various concepts I
+introduced. (such a `Container`, `Plane`, `Physical`) Do not be afraid to
+suggest name changes. Naming is important and I'm already not satisfied with
+the nomenclature I came up with.
+
+## Drawbacks and design limitations
+
+### Unique hierarchy
+
+If we implementation this with a cached navigation tree, it's going to be
+easy to assume accidentally that we have a single fully reachable navigation
+tree. This is not obvious, it's easy to just add the `Container` or `Navigable`
+`Component` to an entity that doesn't have itself a `Container` parent more
+than once.
+
+This needs to be solved, I'm thinking that an error message would be enough in
+this case. But it's worth considering design changes that makes this situation
+a non-issue.
+
+### No concurrent submenus
+
+You have to "build up" reactively the submenus as you navigate through them,
+since **there can only be a single path through the hierarchy**.
+
+We expect that on `FocusChanged`, the developer adds the components needed if
+the focus change leads to a new sub-menu.
+
+I'm implementing first to see how cumbersome this is, and maybe revise the
+design with learning from that.
+
+Maybe it's possible to integrate more than one `Container` within other
+containers. The only motivation for this currently is that it makes it easier to
+implement and reason about the system.
+
+### Inflexible navigation
+
+The developer might want to add arbitrary "jump" relations between various
+menus and buttons. This is not possible with this design. Maybe we could add a
+```rust
+  bridges: HashMap<NavCommand, Entity>,
+```
+field to the `Container` to reroute `NavCommand`s that failed to change focus
+within the menu. I think this might enable cycling references within the
+hierarchy and break useful assumptions.
+
+## Unresolved questions
+
+Building the physical navigation tree seems not-so trivial, I'm not sure how
+feasible it is, although I really think it's not that difficult.
+
+
+## Isn't this overkill?
```

Comment:
Absolutely not. I think this is about as simple as we can get while still allowing something properly robust to complex user needs.

And using two seperate frameworks (one for simple cases, and then an incompatible one for complex ones) is a terrible idea due to the fragmentation and migration pain it induces.